src/evaluators/ForStatement.js

```diff
@@ -268,13 +280,151 @@ function ForBodyEvaluation(
   }
 }
 
+let BailOutWrapperClosureRefVisitor = {
+  ReferencedIdentifier(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.name === "arguments") {
+      state.usesArguments = true;
+    }
+  },
+  ThisExpression(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesThis = true;
+  },
+  "BreakStatement|ContinueStatement"(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.label !== undefined) {
+      state.usesGotoToLabel = true;
+    }
+  },
+  ReturnStatement(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesReturn = true;
+  },
+};
+
+function generateRuntimeForStatement(
+  ast: BabelNodeForStatement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm,
+  labelSet: ?Array<string>
+): AbstractValue {
+  let wrapperFunction = new ECMAScriptSourceFunctionValue(realm);
+  let body = t.blockStatement([ast]);
+  ((body: any): FunctionBodyAstNode).uniqueOrderedTag = realm.functionBodyUniqueTagSeed++;
+  wrapperFunction.$ECMAScriptCode = body;
+  wrapperFunction.$FormalParameters = [];
+  wrapperFunction.$Environment = env;
+  // We need to scan to AST looking for "this", "return", labels and "arguments"
+  let functionInfo = {
+    usesArguments: false,
+    usesThis: false,
+    usesReturn: false,
+    usesGotoToLabel: false,
+  };
+
+  traverse(
+    t.file(t.program([t.expressionStatement(t.functionExpression(null, [], body))])),
+    BailOutWrapperClosureRefVisitor,
+    null,
+    functionInfo
+  );
+  traverse.clearCache();
+
+  if (functionInfo.usesReturn || functionInfo.usesArguments || functionInfo.usesGotoToLabel) {
+    // We do not have support for these yet
+    // TODO: issue number upon creation of issue
+    throw new FatalError("TODO: #? handle more loop bail out cases");
+  }
+  let args = [wrapperFunction];
+
+  if (functionInfo.usesThis) {
+    let thisRef = env.evaluate(t.thisExpression(), strictCode);
+    let thisVal = Environment.GetValue(realm, thisRef);
+    Havoc.value(realm, thisVal);
```

Comment:
This seems a bit like papering over a huge crack in the wall. The loop can modify many other things in a way that is totally opaque to Prepack. Yet, only the this value is havoced. Better than nothing, but not much better.

Reply:
Havocing is transitive. So everything reachable gets havoced.

Reply:
The loop body may access locations that are not reachable from this.

Reply:
What locations are you referring to? The havoc logic recursively goes through every binding and object in the AST of the loop and its body. If something isn't being reached, we need to fix it in the havoc logic. If you mean the loop body is accessing something Prepack doesn't know about, then this isn't safe for pure scope.